src/server/common/profiler.js

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright (c) 2012-2015 S-Core Co., Ltd.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+'use strict';
+
+var logger = require('./log-manager');
+var config = require('./conf-manager').conf;
+var onFinished = require('on-finished');
+var url = require('url');
+var HashMap = require('hashmap').HashMap;
+var db = require('./db');
+var pfdb = require('./pfdb');
+var async = require('async');
+
+
+
+function _log(msg) {
+    logger.debug('[PROF] ', msg);
+}
+
+function _err(msg) {
+    logger.error('[PROF] ', msg);
+}
+
+
+function getip(req) {
+  return req.ip
+    || req._remoteAddress
+    || (req.connection && req.connection.remoteAddress)
+    || undefined;
+}
+
+function getFormattedTime() {
+   return new Date();
+}
+
+function getResponseTime(req, res) {
+    if (!req._startTime) {
+        return null;
+    }
+    var diff = process.hrtime(req._startAt);
+    var ms = diff[0] * 1e3 + diff[1] * 1e-6; 
+    return ms;
+}
+
+
+var storeMgr = (function () {
+})();
+
+//
+// ReqInfo
+//
+
+var ReqInfo = function (instId, url, method) { 
+    this._instId = instId;
+    this._url = url;
+    this._method = method;
+    this._min = null;
+    this._max = null;
+    this._avg = null; 
+    this._failCnt = 0;
+    this._totalCnt = 0;
+
+    this._dirty = false;
+}
+
+ReqInfo.prototype.reset = function () {
+    this._min = null;
+    this._max = null;
+    this._avg = null; 
+    this._failCnt = 0;
+    this._totalCnt = 0;
+
+    this._dirty = false;
+}
+
+ReqInfo.prototype.update = function (responseTime) {
+    var self = this;
+    self._avg = (self._avg === null) ? responseTime : ((self._avg + responseTime) / 2.000); 
```

Comment:
Above code, this._avg means very simple EWMA value (Exponentially Weighted Moving Average) with conffient 0.5. Calculating EWMA correctly is not a simple job. Since requests not periodically repeated, we have to calculate take a 'represntative' value with fixed period (maybe by some timer job) before calculation. I think EWMA is useful as a 'monitoring' metric, like load average (yes, some OS CPU load average value is calculated with EWMA), not suitable for profiling. Profiling is not 'performance monitoring' nor 'service health check'.